rclcpp/include/rclcpp/executor.hpp

```diff
@@ -598,34 +562,58 @@ class Executor
   void
   get_next_timer(AnyExecutable::SharedPtr & any_exec)
   {
-    auto it = guard_condition_handles_.begin();
-    while (it != guard_condition_handles_.end()) {
-      auto timer = get_timer_by_handle(*it);
-      if (timer) {
-        // Find the group for this handle and see if it can be serviced
-        auto group = get_group_by_timer(timer);
-        if (!group) {
-          // Group was not found, meaning the timer is not valid...
-          // Remove it from the ready list and continue looking
-          guard_condition_handles_.erase(it++);
+    for (auto weak_node : weak_nodes_) {
+      auto node = weak_node.lock();
+      if (!node) {
+        continue;
+      }
+      for (auto weak_group : node->callback_groups_) {
+        auto group = weak_group.lock();
+        if (!group || !group->can_be_taken_from_.load()) {
           continue;
         }
-        if (!group->can_be_taken_from_.load()) {
-          // Group is mutually exclusive and is being used, so skip it for now
-          // Leave it to be checked next time, but continue searching
-          ++it;
+        for (auto timer_ref : group->timer_ptrs_) {
+          auto timer = timer_ref.lock();
+          if (timer && timer->check_and_trigger()) {
+            any_exec->timer = timer;
+            any_exec->callback_group = group;
+            node = get_node_by_group(group);
+            return;
+          }
+        }
+      }
+    }
+  }
+
+  std::chrono::nanoseconds
+  get_earliest_timer()
+  {
+    std::chrono::nanoseconds latest = std::chrono::nanoseconds::max();
+    bool timers_empty = true;
+    for (auto weak_node : weak_nodes_) {
+      auto node = weak_node.lock();
+      if (!node) {
+        continue;
+      }
+      for (auto weak_group : node->callback_groups_) {
+        auto group = weak_group.lock();
+        if (!group || !group->can_be_taken_from_.load()) {
           continue;
         }
-        // Otherwise it is safe to set and return the any_exec
-        any_exec->timer = timer;
-        any_exec->callback_group = group;
-        any_exec->node = get_node_by_group(group);
-        guard_condition_handles_.erase(it++);
-        return;
+        for (auto timer_ref : group->timer_ptrs_) {
```

Comment:
Well, you need to consider what the implications are, but yes usually the auto reference is a safe optimization to do. I'd say just update the ones you're adding and anything related to your changes. If you want to audit the rest, I'd do that in a separate pr so we can more easily review each case.